### DISCUSSION TITLE
[Question] Freeze a subtype model when it needs fields in a supertype

### DISCUSSION THREAD

**Discussion by AlexV525** [rrousselGit/freezed#577] [Q&A]
**Context**
We have a base class for models:
```dart
abstract class DataModel {
  const DataModel();

  Json toJson();

  @override
  String toString() => GlobalJsonEncoder.convert(toJson());
}
```

And we have an ORM model for most of subtype models:
```dart
abstract class ORMModel extends DataModel {
  const ORMModel({
    required this.id,
    required this.createTime,
    required this.updateTime,
  });

  final Object id;
  @JsonKey(fromJson: datetimeFromJson, toJson: datetimeToString)
  final DateTime createTime;
  @JsonKey(fromJson: datetimeFromJson, toJson: datetimeToString)
  final DateTime updateTime;

  ValueKey<String> valueKey([String tag = 'default']) {
    return ValueKey<String>(
      '$runtimeType-$id-${updateTime.millisecondsSinceEpoch}-$tag',
    );
  }
}
```

Consider we have a model like:
```dart
@immutable
@JsonSerializable()
class Item extends ORMModel {
  const Item({
    required this.type,
    required Object id,
    required DateTime createTime,
    required DateTime updateTime,
  }) : super(id: id, createTime: createTime, updateTime: updateTime);

  factory Item.fromJson(Json json) => _$ItemFromJson(json);

  final int type;

  @override
  Json toJson() => _$ItemToJson(this);
}
```

How do we manage to make `Item` **freezed**?

**Describe the solution you'd like**
Support to freeze model like above.

**Describe alternatives you've considered**
I know that getters are only supported with the `_` constructor, but with that constructor, we still can't use fields in supertypes.

**Comment by rrousselGit** [comment]
There is no support for super constructors. Consider using implements/with instead of extends

**Reply by AlexV525** [reply]
> Consider using implements/with instead of extends

This seems unimplementable at present, `Item` must extend `DataModel`/`ORMModel` to get through the constructor, which is required by the `fromJson` method.

**Reply by rrousselGit** [reply]
the DataModel you've described could be a mixin. The ORMModel too

```dart
mixin DataModel {
  Json toJson();

  @override
  String toString() => GlobalJsonEncoder.convert(toJson())
}

mixin ORMModel on DataModel {
  Object get id;
  DateTime get createTime;
  DateTime get updateTime;

  ValueKey<String> valueKey([String tag = 'default']) {
    return ValueKey<String>(
      '$runtimeType-$id-${updateTime.millisecondsSinceEpoch}-$tag',
    );
  }
}

@freezed
class Item with DataModel, ORMModel {
  const Item._();
  const factory Item({
    required this.type,
    required Object id,
    @JsonKey(fromJson: datetimeFromJson, toJson: datetimeToString)
     required DateTime createTime,
    @JsonKey(fromJson: datetimeFromJson, toJson: datetimeToString)
    required DateTime updateTime,
  }) = _Item;

  factory Item.fromJson(Json json) => _$ItemFromJson(json);
}
```

**Reply by CrossBread** [reply]
Thanks @rrousselGit , I've been scratching my head trying to sort out exactly how to do this and wondered if `mixin` was the solution, but it wasn't clear from the documentation whether Freezed would play well with fields coming from the mixin.

One further question for clarity, does the `_$Item` mixin go before the others so they also inherit from it, or after the others so that the generated mixin inherits from both `DataModel` and `ORMModel`?

```dart
@freezed
class Item with _$Item, DataModel, ORMModel {/*...*/}
```
Or
```dart
@freezed
class Item with DataModel, ORMModel, _$Item {/*...*/}
```
Thanks!

**Reply by jetaggart** [reply]
@CrossBread any chance you figured this out? Stuck on this same problem.
